amoro-ams/src/main/java/org/apache/amoro/server/optimizing/maintainer/IcebergTableMaintainer.java

```diff
@@ -696,26 +702,152 @@ CloseableIterable<FileEntry> fileScan(
+      }
     } catch (IOException e) {
       throw new RuntimeException(e);
     }
     return expiredFiles;
   }
 
+  private boolean tryExpireByPartition(
+      CloseableIterable<FileEntry> entries,
+      DataExpirationConfig expirationConfig,
+      long expireTimestamp,
+      ExpireFiles expiredFiles) {
+    Types.NestedField expirationField =
+        table.schema().findField(expirationConfig.getExpirationField());
+    Map<Integer, Map<Integer, PartitionField>> expirePartitionFieldsMap =
+        buildExpirePartitionFieldsMap(expirationField);
+    // All historical specs have expirationField as the partition field.
+    boolean allSpecsMatch = expirePartitionFieldsMap.values().stream().noneMatch(Map::isEmpty);
+    if (allSpecsMatch) {
+      Comparable<?> expirePartitionValue;
+      try {
+        expirePartitionValue =
+            getPartitionUpperBound(expirationConfig, expirationField, expireTimestamp);
+      } catch (IllegalArgumentException e) {
+        LOG.error("Failed to get partition upper bound", e);
+        return false;
+      }
+
+      Map<Integer, Map<Integer, Comparable<?>>> expirePartitionValueMap =
+          getExpirePartitionValueMap(
+              expirePartitionFieldsMap, expirationField, expirePartitionValue);
+      entries.forEach(
+          fileEntry -> {
+            List<Boolean> expiredList = new ArrayList<>();
+            ContentFile<?> contentFile = fileEntry.getFile();
+            int fileSpecId = contentFile.specId();
+            for (Map.Entry<Integer, Comparable<?>> entry :
+                expirePartitionValueMap.get(fileSpecId).entrySet()) {
+              Comparable<Object> partitionValue =
+                  contentFile.partition().get(entry.getKey(), entry.getValue().getClass());
+              boolean expired = partitionValue.compareTo(entry.getValue()) < 0;
+              expiredList.add(expired);
+            }
+            if (!expiredList.isEmpty() && expiredList.stream().allMatch(Boolean::booleanValue)) {
+              expiredFiles.addFile(fileEntry);
+            }
+          });
+      return true;
+    }
+    return false;
+  }
+
+  private void expireByMetricsUpperBound(
+      CloseableIterable<FileEntry> entries,
+      DataExpirationConfig expirationConfig,
+      long expireTimestamp,
+      ExpireFiles expiredFiles) {
+    Map<StructLike, DataFileFreshness> partitionFreshness = Maps.newConcurrentMap();
+    Queue<FileEntry> fileEntries = new LinkedTransferQueue<>();
+    entries.forEach(
+        e -> {
+          if (mayExpired(e, partitionFreshness, expireTimestamp)) {
+            fileEntries.add(e);
+          }
+        });
+    fileEntries
+        .parallelStream()
+        .filter(e -> willNotRetain(e, expirationConfig, partitionFreshness))
+        .forEach(expiredFiles::addFile);
+  }
+
+  private Map<Integer, Map<Integer, PartitionField>> buildExpirePartitionFieldsMap(
```

Comment:
Do we need to add some comments to this function here, such as what is the key/value of the return value, and the brief logic